src/user_input.rs

```diff
@@ -1,444 +1,814 @@
+    #[must_use]
+    pub fn button_pressed(&self, button: InputKind) -> bool {
+        match button {
+            InputKind::DualGamepadAxis(axis) => {
+                let axis_pair = self
+                    .get_input_axis_pair(&UserInput::Single(button))
+                    .unwrap();
+                let x = axis_pair.x();
+                let y = axis_pair.y();
+
+                x > axis.x_positive_low
+                    || x < axis.x_negative_low
+                    || y > axis.y_positive_low
+                    || y < axis.y_negative_low
+            }
+            InputKind::SingleGamepadAxis(axis) => {
+                let value = self.get_input_value(&UserInput::Single(button));
+
+                value < axis.negative_low || value > axis.positive_low
+            }
+            InputKind::GamepadButton(gamepad_button) => {
+                // If no gamepad is registered, we know for sure that no match was found
+                if let Some(gamepad) = self.associated_gamepad {
+                    if let Some(gamepad_buttons) = self.gamepad_buttons {
+                        gamepad_buttons.pressed(GamepadButton(gamepad, gamepad_button))
+                    } else {
+                        false
+                    }
+                } else {
+                    false
+                }
+            }
+            InputKind::Keyboard(keycode) => {
+                if let Some(keyboard_stream) = self.keyboard {
+                    keyboard_stream.pressed(keycode)
+                } else {
+                    false
+                }
+            }
+            InputKind::Mouse(mouse_button) => {
+                if let Some(mouse_stream) = self.mouse {
+                    mouse_stream.pressed(mouse_button)
+                } else {
+                    false
+                }
+            }
+        }
+    }
+
+    /// Are all of the `buttons` pressed?
+    #[must_use]
+    pub fn all_buttons_pressed(&self, buttons: &PetitSet<InputKind, 8>) -> bool {
+        for &button in buttons.iter() {
+            // If any of the appropriate inputs failed to match, the action is considered pressed
+            if !self.button_pressed(button) {
+                return false;
+            }
+        }
+        // If none of the inputs failed to match, return true
+        true
+    }
+
+    /// Get the "value" of the input.
+    ///
+    /// For binary inputs such as buttons, this will always be either `0.0` or `1.0`. For analog
+    /// inputs such as axes, this will be the axis value.
+    ///
+    /// [`UserInput::Chord`] inputs are also considered binary and will return `0.0` or `1.0` based
+    /// on whether the chord has been pressed.
+    pub fn get_input_value(&self, input: &UserInput) -> f32 {
+        let use_button_value = || {
+            if self.input_pressed(input) {
+                1.0
+            } else {
+                0.0
+            }
+        };
+
+        if let Some(gamepad) = self.associated_gamepad {
```

Comment:
I suggest refactoring this code so that there is one `match` expression on the outside, and one or more `if` expression(s) dealing with the gamepads inside their respective `match` arm(s).  That way, you won't hit the current problem where you forget to handle non-gamepad related stuff consistently between two `if` branches.

Specifically:
- The top branch of the `if` does not handle `UserInput::Chord`, so you can only get chord values if you have a gamepad attached.
- The bottom branch of the `if` does not handle `UserInput::VirtualDPad`, so you can only get virtual direction pad values if you _don't_ have a gamepad attached.